flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/DebeziumUtils.java

```diff
@@ -233,4 +235,80 @@ private static Map<String, String> querySystemVariables(
 
         return variables;
     }
+
+    public static BinlogOffset findBinlogOffset(long targetMs, MySqlConnection connection) {
+        MySqlConnection.MySqlConnectionConfiguration config = connection.connectionConfig();
+        BinaryLogClient client =
+                new BinaryLogClient(
+                        config.hostname(), config.port(), config.username(), config.password());
+
+        List<String> binlogFiles = new ArrayList<>();
+        JdbcConnection.ResultSetConsumer rsc =
+                rs -> {
+                    while (rs.next()) {
+                        String fileName = rs.getString(1);
+                        long fileSize = rs.getLong(2);
+                        if (fileSize > 0) {
+                            binlogFiles.add(fileName);
+                        }
+                    }
+                };
+
+        try {
+            connection.query("SHOW BINARY LOGS", rsc);
+            LOG.info("Total search binlog: {}", binlogFiles);
+
+            if (binlogFiles.isEmpty()) {
+                return BinlogOffset.ofBinlogFilePosition("", 0);
+            }
+
+            int index = searchBinlogOffset(client, targetMs, binlogFiles);
+            String binlogName = binlogFiles.get(index);
+            return BinlogOffset.ofBinlogFilePosition(binlogName, 0);
+        } catch (Exception e) {
+            throw new FlinkRuntimeException(e);
+        }
+    }
+
+    public static int searchBinlogOffset(
+            BinaryLogClient client, long targetMs, List<String> binlogFiles) {
+        int startIdx = 0;
+        int endIdx = binlogFiles.size() - 1;
+
+        while (startIdx <= endIdx) {
+            int mid = startIdx + (endIdx - startIdx) / 2;
+            FindOffsetListener midBinlog = getBinlogOffsetListener(binlogFiles, mid, client);
+            long midTs = midBinlog.getFirstEventTs();
+            if (midTs < targetMs) {
+                startIdx = mid + 1;
+            } else if (targetMs < midTs) {
+                endIdx = mid - 1;
+            } else {
+                return mid;
+            }
+        }
+
+        return endIdx;
+    }
+
+    public static FindOffsetListener getBinlogOffsetListener(
+            List<String> binlogFiles, int startId, BinaryLogClient client) {
+        String binlogFile = binlogFiles.get(startId);
+        FindOffsetListener findOffsetListener = new FindOffsetListener();
+        client.registerEventListener(findOffsetListener);
+        client.setBinlogFilename(binlogFile);
+        client.setBinlogPosition(0);
+
+        try {
+            LOG.info("begin parse binlog: {}", binlogFile);
+            client.connect();
+        } catch (Throwable e) {
```

Comment:
I think it's better to not use an exception to stop a job normally, we do it through a signal or straightly do it in the event listener.

Reply:
Good advice, after my check，there seems we can not give signal or straightly search all binlogs in one event listener, when event pass in, we have no chance to give signal to outside except `throwable`，event the `exception` is catched by binlog listener, and we could not change the event passed in the event listener.

                if (this.isConnected()) {
                    this.eventLastSeen = System.currentTimeMillis();
                    this.updateGtidSet(event);
                    this.notifyEventListeners(event);  -- sesems only throwable can work
                    this.updateClientBinlogFilenameAndPosition(event);
                }
            }
        } catch (Exception var15) {
            Exception e = var15;

private void notifyEventListeners(Event event) {
        if (event.getData() instanceof EventDeserializer.EventDataWrapper) {
            event = new Event(event.getHeader(), ((EventDeserializer.EventDataWrapper)event.getData()).getExternal());
        }

        Iterator var2 = this.eventListeners.iterator();

        while(var2.hasNext()) {
            EventListener eventListener = (EventListener)var2.next();

            try {
                eventListener.onEvent(event);
            } catch (Exception var5) {
                if (this.logger.isLoggable(Level.WARNING)) {
                    this.logger.log(Level.WARNING, eventListener + " choked on " + event, var5);
                }
            }
        }

    }

Reply:
Maybe we can just call the disconnect method in the listener? I described the method signature and below is the code written by GPT:

---

The `getFirstEventTimestampInBinlog` method needs to fetch the timestamp of the first event in a given binlog file using the `BinaryLogClient`. As of my knowledge cutoff date in 2023, the `BinaryLogClient` from the `mysql-binlog-connector-java` library doesn't provide a direct method to fetch the timestamp of the first event in a binlog file. Therefore, we need to manually connect to that binlog file and read events until we find the first event with a timestamp.

Below is an implementation of the `getFirstEventTimestampInBinlog` method, assuming that you are using the `mysql-binlog-connector-java` library. This implementation requires registering an event listener to the `BinaryLogClient` and reading events until we find the first event with a non-null timestamp.

```java
import com.github.shyiko.mysql.binlog.BinaryLogClient;
import com.github.shyiko.mysql.binlog.event.Event;
import com.github.shyiko.mysql.binlog.event.EventData;
import com.github.shyiko.mysql.binlog.event.EventHeaderV4;
import com.github.shyiko.mysql.binlog.event.RotateEventData;

import java.io.IOException;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

public class BinlogFinder {

    // Other methods...

    private static long getFirstEventTimestampInBinlog(BinaryLogClient client, String binlogFile) throws IOException, InterruptedException {
        final BlockingQueue<Long> timestampQueue = new ArrayBlockingQueue<>(1);

        BinaryLogClient.EventListener eventListener = new BinaryLogClient.EventListener() {
            @Override
            public void onEvent(Event event) {
                EventData data = event.getData();
                if (data instanceof RotateEventData) {
                    // We skip RotateEventData because it does not contain the timestamp we are interested in.
                    return;
                }

                EventHeaderV4 header = event.getHeader();
                long timestamp = header.getTimestamp();
                if (timestamp > 0) {
                    // We've found the first event with a timestamp. Offer it to the queue and stop the client.
                    timestampQueue.offer(timestamp);
                    client.disconnect();
                }
            }
        };

        client.setBinlogFilename(binlogFile);
        client.registerEventListener(eventListener);

        try {
            client.connect();

            // Wait for the timestamp to be available in the queue or until we reach a timeout.
            Long timestamp = timestampQueue.poll(30, TimeUnit.SECONDS);
            if (timestamp != null) {
                return timestamp;
            } else {
                throw new IOException("Failed to retrieve the first event timestamp within the timeout period.");
            }
        } finally {
            client.unregisterEventListener(eventListener);
        }
    }
    
    // Main method and findBinlogFileByTimestamp...
}
```

This code sets up the `BinaryLogClient` to listen for events in the specified binlog file. It uses an `EventListener` to process incoming events. When the first event with a timestamp is found, it is added to a `BlockingQueue` and the client is disconnected. The `getFirstEventTimestampInBinlog` method waits for the timestamp to be available in the queue for up to 30 seconds. If no timestamp is found within the timeout period, it throws an `IOException`.

Please note that interacting with the binary log like this can be complex and may require handling various edge cases and MySQL server configurations. The example above is a simplified implementation and should be tested and modified to suit the specifics of the environment and MySQL server version you are working with.

Reply:
Great, I see.

Reply:
@whhe I have implemented close in Listener when find interest timestamp, pls review again.